sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala

```diff
@@ -837,6 +847,237 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
+      val expGenFunDesc = ExprNodeGenericFuncDesc.newInstance(funInfo.getGenericUDF, concatColumns)
+      exporNodeDesc.add(expGenFunDesc)
+      if (!isRecursive) {
+        // construct genericUDFConcat ExprNode
+        val expConstDesc = new ExprNodeConstantDesc(TypeInfoFactory.stringTypeInfo, value)
+        exporNodeDesc.add(expConstDesc)
+      }
+    } catch {
+      case e: Throwable =>
+        logError(s"conver expression to node failed ${e.getMessage}!")
+        throw e
+    }
+    logDebug("convertExpDescs-exporNodeDesc:" + exporNodeDesc)
+    exporNodeDesc
+  }
+
+  private def convertUDFExpToPartitions(hive: Hive,
+      table: Table,
+      op: BinaryComparison,
+      a: Expression,
+      v: Object): Option[ExprNodeGenericFuncDesc] = {
+    try {
+      val opFun = FunctionRegistry.getFunctionInfo(op.sqlOperator)
+      logDebug("convertUDFExpToPartitions-table:" + table + ",op:" +
+        op + ",a:" + a + ",v:" + v + ",isSubstr")
+      if (opFun.getGenericUDF == null) {
+        throw new UDFArgumentException(s"${op.sqlOperator} is an aggregation " +
+          s"function or a table function.")
+      }
+      Some(ExprNodeGenericFuncDesc.newInstance(
+        opFun.getGenericUDF, convertExpDescs(hive, table, a, v, false)))
+    } catch {
+      case e: Throwable =>
+        logError(s"convert predicates expression failed ${e.getMessage}!")
+        None
+    }
+  }
+
+  private def convertAttributeExpToPartitions(name: String,
+      dataType: DataType, op: BinaryComparison,
+                                              value: Object): Option[ExprNodeGenericFuncDesc] = {
+    val AttributeColumn = new JArrayList[ExprNodeDesc]()
+    AttributeColumn.add(new ExprNodeColumnDesc(
+      TypeInfoFactory.getPrimitiveTypeInfo(dataType.typeName),
+      name, "", false))
+    val expConstDesc = new ExprNodeConstantDesc(TypeInfoFactory.stringTypeInfo, value.toString)
+    AttributeColumn.add(expConstDesc)
+    val funInfo = FunctionRegistry.getFunctionInfo(op.symbol)
+    Some(ExprNodeGenericFuncDesc.newInstance(funInfo.getGenericUDF, AttributeColumn))
+  }
+  def convertUDFFilters(hive: Hive,
+      table: Table, filters: Seq[Expression]): Seq[ExprNodeGenericFuncDesc] = {
+    filters.flatMap(convertExp(hive, table))
+  }
+
+  def convertExp(hive: Hive, table: Table)(expr: Expression): Option[ExprNodeGenericFuncDesc] = {
+    val useConcatAdvancedFilter = SQLConf.get.metastorePredicateConcatFilter
+    logDebug("convertExp-expr:" + expr)
+    expr match {
+      // Prune partition of Concat
+      case op@BinaryComparison(a: Concat, Literal(v, _: StringType)) =>
+        convertUDFExpToPartitions(hive, table, op, a, v.toString)
+      case op@BinaryComparison(Literal(v, _: StringType), a: Concat) =>
+        convertUDFExpToPartitions(hive, table, op, a, v.toString)
+
+      // Prune partition of ConcatWs
+      case op@BinaryComparison(a: ConcatWs, Literal(v, _: StringType)) =>
```

Comment:
Nah .. I don't think this is a right direction in Spark. It's a very big overhead to maintain the list of case Hive function matches, and we should check each behaviour between Spark and Hive.